lib/bolt/plugin/puppetdb.rb

```diff
@@ -15,9 +16,77 @@ def hooks
         ['lookup_targets']
       end
 
+      def warn_missing_fact(certname, fact)
+        @logger.warn("Could not find fact #{fact} for node #{certname}")
+      end
+
+      def fact_name(raw_fact)
+        fact = raw_fact.split(".")
+        fact.drop(1) if fact[0] == 'facts'
+      end
+
       def lookup_targets(opts)
-        nodes = @puppetdb_client.query_certnames(opts['query'])
-        nodes.map { |certname| { 'uri' => certname } }
+        targets = @puppetdb_client.query_certnames(opts['query'])
+        facts = []
+        facts << fact_name(opts['name']) if opts.key?('name')
+        facts << fact_name(opts['uri']) if opts.key?('uri')
+
+        if opts.key?('config')
+          # This is done in parts instead of in place so that we only need to
+          # make one puppetDB query
+          Bolt::Util.walk_vals(opts['config']) do |value|
+            if value.is_a?(String)
+              facts << fact_name(value)
+            end
+            value
+          end
+        end
+
+        facts.uniq!
+        # Returns {'mycertname' => [{'path' => ['nested', 'fact'], 'value' => val'}], ... }
+        fact_values = facts.empty? ? {} : @puppetdb_client.fact_values(targets, facts)
+
+        targets.map do |certname|
+          target_data = fact_values[certname]
+
+          target = {}
```

Comment:
If we build up the target generically first we can then have a simpler check at the end.

target['uri] = certname unless target['uri'] || target['name']